### ISSUE TITLE
Confusing UnimplementedError

### ISSUE THREAD

**Issue by stevenroose** [dart-lang/reflectable#64]
I'm sorry to bother you with all my problems, but I can't seem to solve this one so I'm not sure if it's a bug. I'm trying to port the websockets package to use reflectable, my current WIP is [here](https://github.com/stevenroose/dart-websockets/tree/reflectable).

I got this strange error and I can't seem to find out exactly what it originates from:
```
00:33 +0 -1: test/html_test.dart: connecting and closing                                           
  UnimplementedError
  dart:_internal                                                 JsLibraryMirror.libraryDependencies
  package:reflectable/src/reflectable_mirror_based.dart 2237:14  ReflectableImpl._directlySupportedClasses
  package:reflectable/src/reflectable_mirror_based.dart 2323:36  ReflectableImpl._supportedClasses.<fn>
  dart:_internal                                                 JsLinkedHashMap.2.H.computeSignature.V.func
  package:websockets/src/websocket_base.dart 12:1                J.x
  package:reflectable/src/reflectable_mirror_based.dart 2322:12  ReflectableImpl._supportedClasses
  package:reflectable/src/reflectable_mirror_based.dart 635:15   _LibraryMirrorImpl._rawDeclarations.<fn>
  ===== asynchronous gap ===========================
  dart:_internal                                                 _asyncHelper
  package:websockets/src/websocket.dart 23:1                     <fn>
  ===== asynchronous gap ===========================
  dart:_internal                                                 _wrapJsFunctionForAsync
  package:websockets/src/websocket.dart 23:1                     null.<fn>
  ===== asynchronous gap ===========================
  dart:_internal                                                 _wrapJsFunctionForAsync
  test/html_test.dart 27:1                                       main.<fn>
  package:stack_trace                                            StackZoneSpecification._run
```

It appears as if library dependencies are requested at one point, which my code does not do. The pointers to my code ([`websocket_base.dart`](https://github.com/stevenroose/dart-websockets/blob/reflectable/lib/src/websocket_base.dart#L12) and [`websocket.dart`](https://github.com/stevenroose/dart-websockets/blob/reflectable/lib/src/websocket.dart#L23)) are just the declarations of classes, no interactions with reflectable.

**Comment by stevenroose** [comment]
I'll try to explain what happens.

- The env/html.dart file is imported into the test: https://github.com/stevenroose/dart-websockets/blob/reflectable/test/html_test.dart#L6

- The library is called `websockets.env.html` and it declares one class, HtmlWebSocket: https://github.com/stevenroose/dart-websockets/blob/reflectable/lib/env/html.dart#L10

- `tools.dart` has a `@GlobalQuantifyCapability` for all libraries starting with `websockets.env`: https://github.com/stevenroose/dart-websockets/blob/reflectable/lib/src/environments/tools.dart#L9

- The reflector has following capabilities: `libraryCapability, staticInvokeCapability, declarationsCapability`: https://github.com/stevenroose/dart-websockets/blob/reflectable/lib/src/websocket.dart#L85

- So, it finds the `websockets.env.html` library here: https://github.com/stevenroose/dart-websockets/blob/reflectable/lib/src/environments/tools.dart#L36

- The `WebSocket.connect` tests if this works and we can get in this if block: https://github.com/stevenroose/dart-websockets/blob/reflectable/lib/src/websocket.dart#L85

- Then we try to instantiate a new HtmlWebSocket instance: https://github.com/stevenroose/dart-websockets/blob/reflectable/lib/src/environments/tools.dart#L46

That's somewhere where the error is thrown. The connect method never returns, but the code gets into it.

**Comment by stevenroose** [comment]
Since the `reflectable_mirror_based.dart` class is mentioned, is it possible that the code is not correctly transformed?

These are my transformers:
```yaml

transformers:
- reflectable:
    reflectable/src/transform_import:
      $include:
        - lib/websockets.dart
        - lib/env/html.dart
        - lib/env/io.dart
    entry_points:
      - test/html_test.dart
```

**Comment by eernstg** [comment]
First: Thanks, and please keep these reports coming, everybody benefits from bug fixes!
Will return to the actual topic a bit later today. ;-)

**Comment by eernstg** [comment]
As you mention, the fact that `reflectable_mirror_based.dart` is involved reveals that reflectable is running in untransformed mode.

The transformer `transform_import` mentioned in your 'pubspec.yaml' is a separate transformer (a very simple one) which does the job of switching the 'reflectable.dart' library from using 'dart:mirrors' to using the code generated by the (normal, main) reflectable transformer. It should not be used in any other packages than reflectable itself. So you should be able to use this simplified transformers section:

```yaml
transformers:
- reflectable:
    entry_points:
      - test/html_test.dart
```

However, I suspect that the lines about 'transform_import' in your current 'pubspec.yaml' will end up generating an extra setting for the invocation of the (normal) reflectable transformer, saying that the property 'reflectable/src/transform_import' has been specified for this run of the transformer, and its value is a mapping that maps '$include' to the list ["lib/websockets.dart", "lib/env/html.dart", "lib/env/io.dart"], which is ignored by the (normal) reflectable transformer because it knows nothing about such a setting. ;-)

In other words, that whole thing should not have any effect, and I'm afraid there is something else which prevents the transformation from taking place.

**Comment by stevenroose** [comment]
Yes indeed, the error persists as I removed that part from the pubspec.

Any idea what this is? Can I do something to help resolve it? I use `pub run test -p "chrome"` to test this, so I cannot easily trace or put breakpoints..

**Comment by eernstg** [comment]
I fetched the repo at 'https://github.com/stevenroose/dart-websockets', did `git co reflectable`, and tried to do `pub get`; `pub build --mode=debug test`.

The resulting code could not be translated because it contains code that makes an attempt to access a private class in a different library: `const prefix0._WebSocketsEnvReflector()`.

We should not generate such an expression, so there should be an error during transformation (we catch many other kinds of private access which will fail, but this one apparently slipped through).

However, a private reflector class is not supported (and we haven't had any up to this point, I believe), because the generated code (which will be in a different library than the one that contains the reflector class, here `_WebSocketsEnvReflector`) needs to contain expressions denoting each reflector class.

So please make that class public and try again. ;)

**Comment by stevenroose** [comment]
Weird, the class is private because I use the public instance in the top of the file `@reflector`. But I will try.

Anyhow, I made the class public and pushed it. `pub build --mode=debug test` now runs fine here.

**Comment by eernstg** [comment]
Maybe `pub run test` runs the untransformed code? Can you execute the code in 'build/test/..' directly?

**Comment by stevenroose** [comment]
This is what the console logs:
https://gist.github.com/stevenroose/8eaacdcf19494d539dff

It's not very clear what happens. The print of `startign test` (misspelled, I know) I did put above the `WebSocket.connect` call [here](https://github.com/stevenroose/dart-websockets/blob/reflectable/test/html_test.dart#L14), so that's where it fails, just like with pub test. However it seems to throw different errors..

**Comment by stevenroose** [comment]
It seems as if just nothing happens and the `connect` call is timed out. Even though I have a print right after that call, that is not printed in the console...

**Comment by stevenroose** [comment]
Are you on Freenode perhaps? I'm in the #dart channel, `stevenroose`...

**Comment by eernstg** [comment]
Actually I haven't used Freenode (on top of that, I'll have to concentrate on getting ready for a meeting now --- but I'll be back tomorrow).

However, it looks like the connection to reflectable is gone at this point. Presumably the problem arises in this branch and not in the non-reflectable branch? Otherwise it might not have anything to do with reflectable, and then it would be useful to take it to people who know about all those js_primitives.

**Comment by stevenroose** [comment]
Ok, did some experimenting and found that indeed pub run test doesn't seem to transform code... I reported it here https://github.com/dart-lang/test/issues/383

That being said, I got the first use case working, all tests pass using dart2js.

There are the two use cases that websockets should support (of which the second is preferred):
- The user explicitly imports one of the `websockets/env/*` files to indicate what environment the code is running in. Currently `io` and `html` are supported. The library will than search for libraries with `websockets.env.*`, like I explained [above](https://github.com/dart-lang/reflectable/issues/64#issuecomment-171078511), and use the correct WebSocket implementation. This use case now works fine.
- Some environments, like `io` and `html` can be detected automatically so that the explicit import is not required. We do this by reflecting `dart.(io|html).WebSocket` and see of one of these is present. Then, the right one can be used to create a WebSocket instance.

This second use case did not work using `dart:mirrors` because, unlike `dart.io.WebSocket.connect()` being a static method, in `dart:html` a WebSocket is opened by calling `new dart.html.WebSocket()` and [constructors of native classes are not included in mirrors](https://github.com/dart-lang/sdk/issues/18541).

So now I try to get this second use case working using reflectable. It already works for `dart:io`, but when run in the browser after building, I get a lot of JS-related errors that I have a hard time understanding:
https://gist.github.com/stevenroose/60bf8a317ff4620dafa9

Are you aware of issues when using native constructors with reflectable? Or default constructors (should be `classMirror.newInstance("", [params])`)?

**Comment by stevenroose** [comment]
Aha. Now it comes to this: I print all available libraries (`reflector.libraries`) and this is the output:
```
{reflectable://0/websockets.environments.html.env_html: Instance of 'LibraryMirrorImpl', reflectable://1/websockets.environments.tools: Instance of 'LibraryMirrorImpl', reflectable://2/websockets.environments.exception: Instance of 'LibraryMirrorImpl', reflectable://3/websockets.environments.html.html_websocket_impl: Instance of 'LibraryMirrorImpl', reflectable://4/websockets.environments.io.env_io: Instance of 'LibraryMirrorImpl', reflectable://5/websockets.environments.io.io_websocket_impl: Instance of 'LibraryMirrorImpl'}
```
It seems like native `dart.*` libraries are not included after transformation. Even though I explicitly add [`@GlobalQuantifyCapability(r"^dart.html.WebSocket$", tools.reflector)`](https://github.com/stevenroose/dart-websockets/blob/reflectable/lib/src/environments/html/env_html.dart#L9)

**Comment by stevenroose** [comment]
This thread is getting ridiculous, but I don't care. [I got pub run test to work with transformers](https://github.com/dart-lang/test/issues/383#issuecomment-171365421) and now I got an output I can do something with:
```
00:05 +0: test/html_test.dart: connecting and closing                                              
startign test
reflector: Instance of 'WebSocketsEnvReflector'
searching libraries for websockets.env.*
00:05 +0 -1: test/html_test.dart: connecting and closing                                           
  NullError: method not found: 'get$libraryMirrors' on null
  package:reflectable/src/reflectable_transformer_based.dart 2201:9    ReflectableImpl.libraries
  dart:collection                                                      presentEnv
  package:websockets/src/environments/tools.dart 47:21                 <fn>
  ===== asynchronous gap ===========================
  dart:_internal                                                       _asyncHelper
  package:websockets/src/environments/tools.dart 47:21                 <fn>
  ===== asynchronous gap ===========================
  dart:_internal                                                       _wrapJsFunctionForAsync
  package:websockets/src/environments/tools.dart 47:21                 WebSocket.connect
  ===== asynchronous gap ===========================
  dart:_internal                                                       _wrapJsFunctionForAsync
  http://localhost:8080/html_test_reflectable_original_main.dart 30:1  main.<fn>
  package:stack_trace           
```
It throws on calling `reflector.libraries` [here](https://github.com/stevenroose/dart-websockets/blob/reflectable-debug/lib/src/environments/tools.dart#L43).

**Comment by eernstg** [comment]
Very good, we're getting closer!

When `ReflectableImpl.libraries` gets a null error during the evaluation of `data[this].libraryMirrors`, it must be the case that the "database" of information about the target program (`data`) has no entry for the given reflector (`this`). There are no checks catching this situation in the code because such a thing should never happen.

However, it will happen in a particular situation: Many recent versions of the analyzer exhibit a problem that affects reflectable; the problem is that certain elements are unresolved. For instance, the analyzer reports to the reflectable transformer that a given constant has the value null, not because it is null but because the work that needs to be done in order to compute that value has not been done. We do not have an API to force the resolution to take place (and we cannot see that the problem is a missing resolution step, we just get the null value). As a result, the reflectable database remains empty or near-empty, and in particular it is possible that a reflector has no data in the database at all.

The versions of the analyzer that will work are 0.26.1+14 (and older) and 0.27.1. The versions from 0.26.1+15 and up to 0.27.0 do not work, and neither does 0.27.1+1 and up. We have requested a solution, but none is available yet.

Because of this, older versions of reflectable insisted on analyzer '<= 0.26.1+14', and 0.5.1 insists on analyzer 0.27.1 (just that single version, not a range).

However, reflectable version 0.5.0 allowed for analyzer '^0.27.0' because we noted that 0.27.1 was working, and we did not anticipate that 0.27.1+1 would again exhibit the same problem. So if you are using reflectable 0.5.0 it will allow for some versions of the analyzer that will not work, and an upgrade to use reflectable 0.5.1 should fix it.

Another possible explanation could be 'dependency_overrides' which could force the chosen version of the analyzer to be anything.

One easy step that you could take to clarify the situation would be doing `pub upgrade` and checking which version of the analyzer it uses.

Another thing that could be done in order to see the direct culprit is the following: Make sure 'pubspec.yaml' contains the 'formatted: true' option to the reflectable transformer. Run `pub build --mode=debug test` to generate the transformed files in 'build/test/*.dart'. Check out the resulting reflectable database (it should be in 'build/test/html_test.dart'). If it contains `final _data = {};` then the reflectable database is empty, which almost certainly means that the analyzer version is one of those that exhibit the incomplete resolution problem.

**Comment by stevenroose** [comment]
My `pubspec.lock` file currently has analyzer version 0.27.1. `pub upgrade` says `analyzer 0.27.1 (0.27.1+2 available)`. I for now added an explicit dependency on v0.27.1.

And yes indeed the `_data` map is empty.

**Comment by eernstg** [comment]
OK, the contents of `main` imply that transformation has taken place, and the `_data` value looks as if it has been done using a version of the analyzer that leaves some parts of the program unresolved. 

You already gave two reasons why the analyzer version must be 0.27.1: pubspec.lock and `pub upgrade` output. It's really, really difficult to see how the version could then be anything else. But we have never seen the 'unresolved constants' problem with analyzer 0.27.1. Weird!

When the `_data` is empty I'd expect output from the transformer to include messages like the following:

```
[Info from Reflectable on test_reflectable|test/export_test.dart]:
Ignoring entry point test_reflectable.test.export_test that does not include the class `Reflectable`.
```

Do you have access to the output from the `pub serve` activity, and does it say something like that? The reason why that situation arises is that a certain compile time constant is reported to be null, which is characteristic of analyzer 0.26.1+15 .. 0.27.0 and 0.27.1+1 and up.

**Comment by floitschG** [comment]
@eernstg: Steven posted a link to the sources. Could you please download them and see if you can reproduce?
@stevenroose: is the link in the first post still the most up to date one? thanks.

**Comment by eernstg** [comment]
`git status` says 'Your branch is up-to-date with 'origin/reflectable' in my clone of https://github.com/stevenroose/dart-websockets, and `pub build --mode=debug test` succeeds, and the `_data` value contains data. So we're looking at different behaviors for the transformer when it's being run by `pub build..` vs. when it's being run by `pub serve` in the context of the complex setup mentioned at https://github.com/dart-lang/test#testing-with-barback.

I'll try to create that testing-with-barback setup and see if I can reconstruct the behavior where the database ends up empty.

**Comment by stevenroose** [comment]
Yes. I only added the explicit analyzer dependency, but the pubspec.lock already had that version, so that's only to make sure. So yes, that's the latest.

The refelctable-debug branch adds some printouts for debugging since the debugger doenst work with dart2js.

**Comment by stevenroose** [comment]
@eernstg With me the `_data` is empty... Strange..

**Comment by stevenroose** [comment]
@eernstg 
```dart
$ pub serve
Loading source assets... 
Loading reflectable and test/pub_serve transformers... 
Serving websockets test on http://localhost:8080
Build completed successfully
```

**Comment by eernstg** [comment]
OK, it wasn't so complex. ;-) But the database is not empty, and I don't get any complaints from the test:

```
...$ pub run test:test --pub-serve=8080 -p chrome                                                                                                                                                                                                                     [12:47]
00:01 +0: compiling test/html_test.dart
"pub serve" is compiling test/html_test.dart...
00:19 +2: All tests passed!
```

and

```
main() {
  _initializeReflectable();
  return original.main();
}

final _data = {const prefix0.WebSocketsEnvReflector(): ...} 
...
```

**Comment by stevenroose** [comment]
Wow.. Euhm, I still got a problem though. Weird.

In my test, it can't find `dart.html`. This is the `reflector.libraries`: https://gist.github.com/stevenroose/864ddd38617bdf72bcdb

There is dart.dom, but no dart.html.. It's strange that the test works for you. Does the `html_test.dart` import "env/html.dart"? It shouldn't..

**Comment by stevenroose** [comment]
Is it possible that you are using the dev version of the SDK?
```
$ dart --version
Dart VM version: 1.13.2 (Tue Jan  5 06:49:57 2016) on "linux_x64"
$ pub --version
Pub 1.13.2```

**Comment by eernstg** [comment]
Arghh, I shouldn't have trusted the 'up-to-date' message:

```
...$ git co reflectable
Switched to branch 'reflectable'
Your branch is up-to-date with 'origin/reflectable'.
...$ git pull --rebase origin reflectable
remote: Counting objects: 9, done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 9 (delta 5), reused 7 (delta 3), pack-reused 0
Unpacking objects: 100% (9/9), done.
From https://github.com/stevenroose/dart-websockets
 * branch            reflectable -> FETCH_HEAD
   a0681a4..b31ed13  reflectable -> origin/reflectable
First, rewinding head to replay your work on top of it...
Fast-forwarded reflectable to b31ed1358544d7200c9206473ffd229b63199724.
```

Re-running, will be back soon.

**Comment by stevenroose** [comment]
Check the import on the top of `html_test.dart`. I don't remember if I got it working with the import already, but it should now work both with and without the import. I should make separate tests, really.

The case without the import is the most important one though.

**Comment by eernstg** [comment]
But `pub get`, `pub serve`, still leads to

```
...$ pub run test:test --pub-serve=8080 -p chrome
00:01 +0: compiling test/html_test.dart                                                            
"pub serve" is compiling test/html_test.dart...
00:19 +2: All tests passed!                   
```

and 

```dart
main() {
  _initializeReflectable();
  return original.main();
}

final _data = {const prefix0.WebSocketsEnvReflector(): ..} ...
```

And yes, I'm using a dev sdk, (dart-sdk-1.14.0-dev.5.0 currently, switching among dev.[4-7].0 for various reasons).

The `import "package:websockets/env/html.dart";` is still present. Commenting it out we get the error (HA!) you mention:

```
...$ pub run test:test --pub-serve=8080 -p chrome
00:01 +0: compiling test/html_test.dart                                                            
"pub serve" is compiling test/html_test.dart...
00:18 +0 -1: test/html_test.dart: connecting and closing                                           
  NullError: method not found: 'get$libraryMirrors' on null
  package:reflectable/src/reflectable_transformer_based.dart 2201:9    ReflectableImpl.libraries
  ...
```

**Comment by stevenroose** [comment]
Yes, now we are there. I forgot the `--pub-serve` flag my last runs. So yes, we are synced now. The only case left not working is the one without the import.

**Comment by stevenroose** [comment]
What happens is looking for `dart.html.WebSocket` in order to instantiate it.

- https://github.com/stevenroose/dart-websockets/blob/reflectable-debug/lib/src/environments/html/env_html.dart#L22
- https://github.com/stevenroose/dart-websockets/blob/reflectable-debug/lib/src/environments/tools.dart#L22

**Comment by eernstg** [comment]
Definitely it should be two distinct tests. ;-)

But in this case we _do_ get an empty database with `pub build --mode=debug test`, and pubspec.lock says that the analyzer is 0.27.1.

That's new!

Looking into it.

**Comment by stevenroose** [comment]
So I created separate test files, see reflectable branch. 

Now I seem to get an error for the case with the env import too. I did update the entry_points though (is wildcarding entry_points possible btw?)..
https://gist.github.com/stevenroose/3a84b521d9cea32b9fd2

**Comment by eernstg** [comment]
We don't currently interpret any kind of wildcards in the entry_points.

**Comment by stevenroose** [comment]
Stupid typo in the entry_points.

Now even the dart:io tests are failing:
```
Unsupported operation: The transformed code is running with the untransformed reflectable package. Remember to set your package-root to 'build/.../packages'.
```

Hehe, I'm getting enough of this and I hate to bother you so much with it.

**Comment by stevenroose** [comment]
You can checkout the reflectable branch yourself. The only _real_ bug there is, is still persistent in the dart:html case without the env import (the last test). Both dart:io cases fail too now, but that's due to something I don't understand most probably.

**Comment by eernstg** [comment]
I get the same: 'html_without_env_test.dart' and 'io_without_env_test.dart' have no _data.

PS: It's frustrating, but I want to discover how this situation differs from the known & tested ones, such that we can get an error that was otherwise associated with versions of the analyzer different from 0.27.1.

**Comment by eernstg** [comment]
I've discovered that, for once, the empty database is not due to unexpected `null` constants from the analyzer, it's actually because there is nothing to match!

For instance, 'io_without_env_test.dart' (indirectly) imports lib/src/environments/io/env_io.dart, lib/src/environments/html/env_html.dart, and lib/src/environments/tools.dart, which means that the following global quantifiers will be present in the program:

```dart
@GlobalQuantifyCapability(r"^dart\.io\.WebSocket$", tools.reflector)
@GlobalQuantifyCapability(r"^dart\.dom\.html\.WebSocket$", tools.reflector)
@GlobalQuantifyCapability(r"^websockets\.env\..*$", reflector)
```

(I adjusted the third one to include the `dom` part while tracing down the issue.) The program never imports 'dart:io' and never imports 'dart:html', so nothing matches r"^dart\.io\.WebSocket$" and nothing matches r"^dart\.dom\.html\.WebSocket$". Similarly, nothing in the program imports lib/env/html.dart nor lib/env/io.dart, so nothing matches r"^websockets\.env\..*$".

In contrast, 'io_with_env_test.dart' imports the needed libraries such that the expected matches are obtained, and similarly for 'html_with_env_test.dart'.

But the tests where we see a failure there are no matches. When a reflector is present but nothing matches, we do not create an entry for it. We will fix that, but this only means that the error message will be more informative, there will still not be anything to match.

The point is that your program never imports the classes that you are searching for, so there is no way this could be made to work.

Having looked a bit more closely at these tests, it looks like you are trying to achieve the effect of a configurable import (check it out [here](https://github.com/dart-lang/dart_enhancement_proposals/blob/master/Meetings/2015-03-04%20DEP%20Committee%20Meeting.md)). This is not something you can emulate by adding a lot of complex import connections, but I believe that you would be able to achieve what you want (with the '*_without_env_test.dart' tests) using almost any of the proposals for such a mechanism. You just can't do it now.

In summary, I'll make the reflectable transformer respond more gracefully to the case where a reflector does not match anything at all, and I'll close this issue.

**Comment by eernstg** [comment]
There is a CL at https://codereview.chromium.org/1602643002/ which adds the checks needed in order to detect a totally unused reflector, such that we get a proper error message rather than the null error that was reported in this issue. The fix will soon be published as part of the upcoming version 0.5.2.

**Comment by stevenroose** [comment]
I'm back. You say 
>The program never imports 'dart:io' and never imports 'dart:html', so nothing matches r"^dart.io.WebSocket$" and nothing matches r"^dart.dom.html.WebSocket$".

This thing is, with dart:mirrors, IIRC all available `dart.*` libraries are available by default, imported or not (The http package from @nex3 makes use of that property). It seems that is not the case with reflectable. 

But good that you got it solved, thanks!

**Comment by stevenroose** [comment]
Also, why does reflectable map `dart:html` to `dart.dom.html` while `dart:mirrors` maps it to `dart.html`?

**Comment by eernstg** [comment]
About having all 'dart:*' libraries automatically if it's on the relevant platform: That would never be true for all platforms where reflectable is used because that would be too expensive.

I think the vm may be special because it needs to do many of the things that some 'dart:*' libraries do in order to work at all (e.g., it may need to print to stdout/stderr as part of the startup even in cases where the Dart program does not import 'dart:io'), and in general the vm is expected to run in lots of memory (there is no tree-shaking, etc.). But even for the vm there may be future scenarios where resources are more limited, and gratuitously importing libraries like 'dart:profiler', 'dart:async', or 'dart:math' would then probably be avoided if at all possible.

So even if such a thing currently works on the vm, I don't think it's a technique that is likely to keep working, or even that it could work on all platforms today.

About the name of 'dart:html': Reflectable uses the names that the analyzer reports, and indeed the library directive in both '$SDK/lib/html/dartium/html_dartium.dart' and '$SDK/lib/html/dart2js/html_dart2js.dart' is `library dart.dom.html;`. So this is the same approach as that which is used for any other library (except for the fact that there is a special-case treatment that makes this particular library configurable, it's just lucky that all the potentially chosen variants have the same name). I'm not sure how 'dart:mirrors' made the choice to call it 'dart.html'.

**Comment by stevenroose** [comment]
Yes. It all works now! 

But you are right, I'm exactly trying to do what a configured import would do. But it works now. `WebSocket.connect` can be called both in the browser and the VM as long as the main program imports `dart:html` or `dart:io`. So the http package should be able to do the same thing.

**Comment by eernstg** [comment]
Cool, thanks!
